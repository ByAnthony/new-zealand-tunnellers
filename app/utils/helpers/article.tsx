import {
  ArticleReferenceData,
  ImageData,
  SectionData,
} from "../../types/article";

export const getSections = (sections: SectionData[]) => {
  return sections.map((section: SectionData) => ({
    title: section.title,
    text: section.text,
  }));
};

export const getImages = (images: ImageData[]) => {
  return images.map((image: ImageData) => ({
    file: image.file,
    title: image.title,
    photographer: image.photographer,
    reference: image.reference,
    alt: image.alt,
  }));
};

export const getNextChapter = (
  chapter: number,
  articles: ArticleReferenceData[],
) => {
  const index = articles.findIndex(
    (article: ArticleReferenceData) => article.chapter === chapter,
  );

  if (index !== -1 && index + 1 < articles.length) {
    const nextArticle = articles[index + 1];
    return {
      url: nextArticle.id,
      chapter: nextArticle.chapter,
      title: nextArticle.title,
    };
  }

  return null;
};

export const formatText = (text: string) => {
  const paragraphs = text.split("\\n\\n");

  let segmentKey: number = 0;

  return paragraphs.map((paragraph) => {
    const segments = paragraph
      .replace(/--/g, "\u00A0")
      .split(/(\*.+?\*)|(\[.+?\))/g);

    const formattedSegments = segments.map((segment) => {
      segmentKey += 1;
      if (segment && segment.startsWith("*") && segment.endsWith("*")) {
        const italicText = segment.slice(1, -1);
        return <em key={segmentKey}>{italicText}</em>;
      }
      if (segment && segment.startsWith("[") && segment.endsWith(")")) {
        const linkText = segment.slice(1, -1);

        const [label, url] = linkText.split("](");

        if (url.includes("footnote")) {
          return (
            <a key={segmentKey} href={url} id={`reference_${label}`}>
              {`[${label}]`}
            </a>
          );
        }
        if (url.includes("reference")) {
          return (
            <a
              key={segmentKey}
              href={url}
              id={`footnote_${label.slice(0, -1)}`}
            >
              {label}
            </a>
          );
        }
        return (
          <a key={segmentKey} href={url}>
            {label}
          </a>
        );
      }
      return <span key={segmentKey}>{segment}</span>;
    });

    return <p key={paragraphs.indexOf(paragraph)}>{formattedSegments}</p>;
  });
};
