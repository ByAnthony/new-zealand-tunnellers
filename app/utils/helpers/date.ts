import { DateObj } from "../../types/tunneller";

export const today = new Date();

export const getYear = (date: string) => {
  return date.slice(0, 4);
};

export const getDayMonth = (date: string) => {
  const datetime = new Date(date);
  const day = datetime.getDate();
  const month = datetime.toLocaleString("default", { month: "long" });

  return `${day} ${month}`;
};

export const getDate = (date: string) => {
  const dateObj: DateObj = { year: getYear(date), dayMonth: getDayMonth(date) };
  return dateObj;
};
