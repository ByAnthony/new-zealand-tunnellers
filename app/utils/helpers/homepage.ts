import { historyChapter, historyImageChapters } from "../../types/homepage";

export const getHistoryChapters = (
  chapters: historyChapter[],
  images: historyImageChapters[],
) => {
  const getChapterImage = (
    images: historyImageChapters[],
    position: number,
  ) => {
    if (position >= 0 && position < images.length) {
      return images[position].file;
    }
    return "";
  };

  return chapters.map((chapter: historyChapter, index: number) => ({
    ...chapter,
    image: getChapterImage(images, index),
  }));
};
