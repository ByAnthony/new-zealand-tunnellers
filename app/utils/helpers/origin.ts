export const getParent = (name: string | null, origin: string | null) => {
  return name ? { name, origin } : null;
};

export const getNzResident = (
  month: number | null,
  enlistment: string | null,
  posted: string | null,
) => {
  if (month) {
    const calculateResidentSince = (year: string, month: number) => {
      if (month < 12) {
        return year;
      }
      const residenceInYear = month / 12;
      const residentSince = Number(year) - residenceInYear;
      return residentSince.toString();
    };

    if (enlistment) {
      const enlistmentYear = enlistment.slice(0, 4);
      return calculateResidentSince(enlistmentYear, month);
    }

    if (posted) {
      const postedYear = posted.slice(0, 4);
      return calculateResidentSince(postedYear, month);
    }
  }
  return null;
};
