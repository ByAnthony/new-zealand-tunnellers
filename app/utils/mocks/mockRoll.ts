import { Tunneller } from '@/app/types/roll';
import { mockId } from './mockProfile';

const johnDoe = {
  forename: 'John',
  surname: 'Doe',
};

const alexanderDriscott = {
  forename: 'Alexander',
  surname: 'Driscott',
};

const williamRight = {
  forename: 'William',
  surname: 'Right',
};

const robertLang = {
  forename: 'Robert',
  surname: 'Lang',
};

export const mockDetails: Tunneller = {
  id: mockId,
  name: johnDoe,
  birthDate: '1886',
  deathDate: '1952',
};

export const mockDetailsList: Tunneller[] = [
  mockDetails,
  {
    ...mockDetails,
    id: 1,
    name: alexanderDriscott,
  },
];

export const mockRoll: Record<string, Tunneller[]> = {
  D: [
    mockDetails,
    {
      ...mockDetails,
      id: 1,
      name: alexanderDriscott,
    },
  ],
  L: [
    {
      ...mockDetails,
      id: 2,
      name: robertLang,
    },
  ],
  R: [
    {
      ...mockDetails,
      id: 3,
      name: williamRight,
    },
  ],
};
