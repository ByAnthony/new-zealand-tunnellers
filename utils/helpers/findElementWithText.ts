import { screen } from "@testing-library/react";

export const findElementWithText = (text: string) => {
  const elements = screen.getAllByText((content) => content.includes(text));

  if (elements.length === 0) {
    throw new Error(`Unable to find an element with the text: ${text}`);
  }

  return elements[0];
};
