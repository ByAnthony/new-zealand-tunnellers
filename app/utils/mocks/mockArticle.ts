import { AboutUsArticle } from "../../types/article";

export const mockArticle: AboutUsArticle = {
  id: "about-us",
  title: "Our Story",
  section: [
    { title: "Title 1", text: "This is the first paragraph of our story." },
    { title: "Title 2", text: "This is the second paragraph of our story." },
  ],
  image: [
    {
      file: "our-story.jpg",
      title: null,
      photographer: null,
      reference: null,
      alt: "A picture of our team",
    },
  ],
};
