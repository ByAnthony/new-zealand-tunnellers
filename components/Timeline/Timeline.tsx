"use client";

import Link from "next/link";

import { today } from "@/utils/helpers/date";

import { Title } from "../Title/Title";
import { HowToCite } from "../HowToCite/HowToCite";
import { TimelineEvents } from "./TimelineEvents/TimelineEvents";
import { TunnellerProfile } from "@/types/tunneller";

import STYLES from "./Timeline.module.scss";

type Props = {
  tunneller: TunnellerProfile;
};

export function Timeline({ tunneller }: Props) {
  return (
    <div className={STYLES.timeline}>
      <div className={STYLES.header}>
        <div className={STYLES.link}>
          <Link href="/tunnellers">Tunnellers</Link>
          <span>/</span>
          <a
            href={`/tunnellers/${tunneller.id}`}
          >{`${tunneller.summary.name.forename} ${tunneller.summary.name.surname}`}</a>
        </div>
        <div className={STYLES["main-title"]}>
          <Title title={"World War I\\Timeline"} />
        </div>
      </div>
      <div className={STYLES.events}>
        <div className={STYLES.line}>
          <TimelineEvents
            militaryYears={tunneller.militaryYears}
            death={tunneller.death}
          />
        </div>
      </div>
      <HowToCite
        id={tunneller.id}
        summary={tunneller.summary}
        today={today}
        timeline
      />
    </div>
  );
}
