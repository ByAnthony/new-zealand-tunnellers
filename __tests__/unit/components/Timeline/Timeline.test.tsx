import { render, screen } from "@testing-library/react";

import { Timeline } from "@/components/Timeline/Timeline";
import { findElementWithText } from "@/utils/helpers/findElementWithText";
import { mockTunnellerProfile } from "@/utils/mocks/mockTunneller";

jest.useFakeTimers().setSystemTime(new Date("2023-05-04"));

describe("Timeline", () => {
  test("matches the snapshot", () => {
    const { asFragment } = render(
      <Timeline tunneller={mockTunnellerProfile} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the component correctly", () => {
    render(<Timeline tunneller={mockTunnellerProfile} />);

    const tunnellersLink = screen.getByText("Tunnellers");
    expect(tunnellersLink).toHaveAttribute("href", "/tunnellers");
    const tunnellerLink = screen.getByText("John Smith");
    expect(tunnellerLink).toHaveAttribute("href", "/tunnellers/1");

    const titleLineOne = screen.getByText("World War I");
    expect(titleLineOne).toHaveClass("title-line-1");
    const titleLineTwo = screen.getByText("Timeline");
    expect(titleLineTwo).toHaveClass("title-line-2");

    // Timeline
    expect(
      screen.getByRole("heading", { name: "Year 1915" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Year 1916" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Year 1917" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Year 1918" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Year 1919" }),
    ).toBeInTheDocument();

    // How to cite
    expect(
      screen.getByRole("heading", { name: "How to cite this page" }),
    ).toBeInTheDocument();

    expect(
      findElementWithText("“World War I Timeline of John Smith“"),
    ).toBeInTheDocument();
    expect(screen.getByText(/Accessed: 4 May 2023/)).toBeInTheDocument();
    expect(
      screen.getByText(
        /URL: www.nztunnellers.com\/tunnellers\/1\/wwi-timeline./,
      ),
    ).toBeInTheDocument();
  });
});
