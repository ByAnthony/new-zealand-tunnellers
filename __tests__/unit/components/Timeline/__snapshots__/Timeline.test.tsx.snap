// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="timeline"
  >
    <div
      class="header"
    >
      <div
        class="link"
      >
        <a
          href="/tunnellers"
        >
          Tunnellers
        </a>
        <span>
          /
        </span>
        <a
          href="/tunnellers/1"
        >
          John Smith
        </a>
      </div>
      <div
        class="main-title"
      >
        <div
          class="main-title"
        >
          <h1>
            <span
              class="title-line-1"
            >
              World War I
            </span>
            <span
              class="title-line-2"
            >
              Timeline
            </span>
          </h1>
        </div>
      </div>
    </div>
    <div
      class="events"
    >
      <div
        class="line"
      >
        <div>
          <h2
            aria-label="Year 1915"
            class="year-title"
          >
            1915
          </h2>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1915
              </div>
              <div
                class="day-month"
              >
                26 January
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="main-event"
              >
                <p>
                  Enlisted
                </p>
                <span>
                  Main Body
                </span>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1915
              </div>
              <div
                class="day-month"
              >
                14 February
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="tunneller-event-with-title"
              >
                <p>
                  Trained
                </p>
                <span>
                  Military Camp, Somewhere
                </span>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1915
              </div>
              <div
                class="day-month"
              >
                4 May
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="main-event"
              >
                <p>
                  Major Event
                </p>
                <span>
                  Something happened
                </span>
              </div>
            </div>
          </div>
        </div>
        <div>
          <h2
            aria-label="Year 1916"
            class="year-title"
          >
            1916
          </h2>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1916
              </div>
              <div
                class="day-month"
              >
                19 November
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="tunneller-event"
              >
                <p>
                  Something happened that day
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1916
              </div>
              <div
                class="day-month"
              >
                12 December
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div>
                <div
                  class="company-event"
                >
                  <img
                    alt="Company event image"
                    class="event-image"
                    data-nimg="1"
                    decoding="async"
                    height="489"
                    loading="lazy"
                    src="/_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=1920&q=75"
                    srcset="/_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=750&q=75 1x, /_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=1920&q=75 2x"
                    style="color: transparent;"
                    width="670"
                  />
                  <p>
                    Something happened that day for the Company
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div>
          <h2
            aria-label="Year 1917"
            class="year-title"
          >
            1917
          </h2>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1917
              </div>
              <div
                class="day-month"
              >
                9 April
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="main-event"
              >
                <p>
                  Major Event
                </p>
                <span>
                  Something happened
                </span>
              </div>
              <div>
                <div
                  class="company-event"
                >
                  <img
                    alt="Company event image"
                    class="event-image"
                    data-nimg="1"
                    decoding="async"
                    height="489"
                    loading="lazy"
                    src="/_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=1920&q=75"
                    srcset="/_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=750&q=75 1x, /_next/image?url=%2Fimages%2Froll%2Fimage.jpg&w=1920&q=75 2x"
                    style="color: transparent;"
                    width="670"
                  />
                  <p>
                    Something happened that day for the Company
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1917
              </div>
              <div
                class="day-month"
              >
                8 August
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="tunneller-event"
              >
                <p>
                  Something happened that day
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <h2
            aria-label="Year 1918"
            class="year-title"
          >
            1918
          </h2>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1918
              </div>
              <div
                class="day-month"
              >
                3 March
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="main-event"
              >
                <p>
                  Major Event
                </p>
                <span>
                  Something happened
                </span>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1918
              </div>
              <div
                class="day-month"
              >
                13 October
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="tunneller-event"
              >
                <p>
                  Something happened that day
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <h2
            aria-label="Year 1919"
            class="year-title"
          >
            1919
          </h2>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1919
              </div>
              <div
                class="day-month"
              >
                4 April
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="main-event"
              >
                <p>
                  Major Event
                </p>
                <span>
                  Something happened
                </span>
              </div>
            </div>
          </div>
          <div>
            <div
              class="date"
            >
              <div
                class="year"
              >
                1919
              </div>
              <div
                class="day-month"
              >
                12 July
              </div>
            </div>
            <div
              class="event-wrapper"
            >
              <div
                class="tunneller-event"
              >
                <p>
                  Something happened that day
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="howtocite"
    >
      <h3>
        How to cite this page
        <button
          class="copy-paste"
        >
          <img
            alt="Copy to clipboard"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/_next/image?url=%2Fcopy.png&w=32&q=75"
            srcset="/_next/image?url=%2Fcopy.png&w=16&q=75 1x, /_next/image?url=%2Fcopy.png&w=32&q=75 2x"
            style="color: transparent;"
            width="13"
          />
        </button>
      </h3>
      <p>
        Anthony Byledbal, “World War I Timeline of John Smith“,
        <em>
           New Zealand Tunnellers Website
        </em>
        , 2023 (2009), Accessed: 4 May 2023. 
        <span>
          URL: www.
          <wbr />
          nztunnellers
          <wbr />
          .com/
          <wbr />
          tunnellers/
          <wbr />
          1/
          <wbr />
          wwi-
          <wbr />
          timeline
        </span>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
